File: cpp/2023/day10.cpp
#include "../aocHelper.h"

#include <cassert>

class Day10 : public BaseDay
{
public:
	Day10() : BaseDay("10") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		std::vector<std::string_view> grid{};

		const char* lineStart = input;

		while (*input != '\0')
		{
			if (*input == '\n')
			{
				grid.push_back({lineStart, input});
				input++;
				lineStart = input;
			}
			else
			{
				input++;
			}
		}

		const int height = grid.size();
		const int width = grid[0].size();

		const auto find_start = [&grid, &width, &height]() -> uint16_t
		{
			for (int y = 0; y < height; y++)
			{
				for (int x = 0; x < width; x++)
				{
					if (grid[y][x] == 'S')
					{
						return x << 8 | y;
					}
				}
			}
			return 0;
		};

		const uint16_t start = find_start();

		const auto get_delta = [](char c) -> std::pair<std::pair<int, int>, std::pair<int, int>>
		{
			switch (c)
			{
				case '|':
				{
					return {{0, -1}, {0, 1}};
				}
				case '-':
				{
					return {{-1, 0}, {1, 0}};
				}
				case 'L':
				{
					return {{0, -1}, {1, 0}};
				}
				case 'J':
				{
					return {{0, -1}, {-1, 0}};
				}
				case '7':
				{
					return {{-1, 0}, {0, 1}};
				}
				case 'F':
				{
					return {{1, 0}, {0, 1}};
				}
				default:
				{
					return {};
				}
			}
		};

		const auto next_pos = [&grid, &get_delta](uint16_t curr, uint16_t prev) -> uint16_t
		{
			int cx = curr >> 8;
			int cy = curr & 0xFF;

			int px = prev >> 8;
			int py = prev & 0xFF;

			int dx = px - cx;
			int dy = py - cy;

			char curr_pos = grid[cy][cx];

			std::pair<int, int> current_delta{dx, dy};

			auto deltas = get_delta(curr_pos);

			std::pair<int, int> other_delta = (deltas.first == current_delta) ? deltas.second : deltas.first;

			int nx = cx + other_delta.first;
			int ny = cy + other_delta.second;
			return nx << 8 | ny;
		};

		const auto apply_adj = [](uint16_t pos, const std::pair<int, int>& adj) -> uint16_t
		{
			int x = (pos >> 8) + adj.first;
			int y = (pos & 0xFF) + adj.second;
			if (x < 0 || y < 0)
			{
				assert(false && __FUNCTION__ ": result is negative");
			}
			return x << 8 | y;
		};

		// part 1

		std::vector<uint16_t> loop{};

		loop.push_back(start);

		{
			uint16_t pos = start;

			int cx = pos >> 8;
			int cy = pos & 0xFF;

			char up = grid[cy - 1][cx];
			char down = grid[cy + 1][cx];
			char left = grid[cy][cx - 1];
			char right = grid[cy][cx + 1];

			if (up == '|' || up == '7' || up == 'F')
			{
				cy--;
			}
			else if (down == '|' || down == 'L' || down == 'J')
			{
				cy++;
			}
			else if (left == '-' || left == 'L' || left == 'F')
			{
				cx--;
			}
			else if (right == '-' || right == 'J' || right == '7')
			{
				cx++;
			}
			else
			{
				assert(false && "No start");
			}

			pos = cx << 8 | cy;

			loop.push_back(pos);

			while (pos != start)
			{
				pos = next_pos(loop[loop.size() - 1], loop[loop.size() - 2]);
				loop.push_back(pos);
			}

			assert(loop.size() % 2 == 1 && "Loop size should be even");
			part1 = loop.size() / 2;
		}

		// part 2

		std::unordered_set<uint16_t> loop_set{loop.begin(), loop.end()};

		const auto bfs = [&grid, &width, &height, &loop_set](uint16_t start) -> std::unordered_set<uint16_t>
		{
			std::deque<uint16_t> to_check{};
			to_check.push_back(start);
			std::unordered_set<uint16_t> seen{};
			seen.insert(start);

			while (!to_check.empty())
			{
				uint16_t curr = to_check.front();
				to_check.pop_front();

				static constexpr std::array<std::pair<int, int>, 4> adj_values{{{0, -1}, {0, 1}, {-1, 0}, {1, 0}}};

				int cx = curr >> 8;
				int cy = curr & 0xFF;

				for (auto adj : adj_values)
				{
					int nx = cx + adj.first;
					int ny = cy + adj.second;
					if (nx >= 0 && nx < width && ny >= 0 && ny < height)
					{
						uint16_t next_pos = nx << 8 | ny;
						if (!loop_set.contains(next_pos))
						{
							if (!seen.contains(next_pos))
							{
								seen.insert(next_pos);
								to_check.push_back(next_pos);
							}
						}
					}
				}
			}

			return seen;
		};

		const auto is_point_in_path = [](uint16_t pos, const std::vector<uint16_t>& poly) -> bool
		{
			// https://en.m.wikipedia.org/wiki/Even%E2%80%93odd_rule
			int size = poly.size();
			int j = size - 1;
			bool c = false;

			int x = pos >> 8;
			int y = pos & 0xFF;

			for (int i = 0; i < size; i++)
			{
				int ix = poly[i] >> 8;
				int iy = poly[i] & 0xFF;
				int jx = poly[j] >> 8;
				int jy = poly[j] & 0xFF;

				if (x == ix && y == iy)
				{
					// point is in corner
					return true;
				}

				if ((iy > y) != (jy > y))
				{
					int slope = (x - ix) * (jy - iy) - (jx - ix) * (y - iy);

					if (slope == 0)
					{
						// point is on boundary
						return true;
					}
					if ((slope < 0) != (jy < iy))
					{
						c = !c;
					}
				}

				j = i;
			}

			return c;
		};

		const auto check_group =
			[&width, &height, &is_point_in_path, &loop](const std::unordered_set<uint16_t>& group) -> bool
		{
			// check touching edges

			for (auto& pos : group)
			{
				int x = pos >> 8;
				int y = pos & 0xFF;
				if (x == 0 || x == width - 1)
				{
					return false;
				}
				if (y == 0 || y == height - 1)
				{
					return false;
				}
			}

			// group is now surrounded by loop

			uint16_t test_pos = *group.begin();

			return is_point_in_path(test_pos, loop);
		};

		{
			std::unordered_set<uint16_t> seen{};

			for (int y = 0; y < height; y++)
			{
				for (int x = 0; x < width; x++)
				{
					uint16_t pos = x << 8 | y;
					if (!seen.contains(pos) && !loop_set.contains(pos))
					{
						auto group = bfs(pos);
						seen.insert(group.begin(), group.end());

						if (check_group(group))
						{
							part2 += group.size();
						}
					}
				}
			}
		}

		return {part1, part2};
	}
};

File: 2023/day10.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day10.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import *

data = parsefile(file_name,  ["\n"])

width = len(data[0])
height = len(data)
#print("grid", width, height, width*height)

start = None
for x in range(width):
	for y in range(height):
		if data[y][x] == "S":
			start = (x, y)
			break
	if start is not None:
		break

deltas = {
	"|": ((0, -1), (0, 1)),
	"-": ((-1, 0), (1, 0)),
	"L": ((0, -1), (1, 0)),
	"J": ((0, -1), (-1, 0)),
	"7": ((-1, 0), (0, 1)),
	"F": ((1, 0), (0, 1))
}


def next_pos(curr, prev):
	#print(curr, prev)
	delta = (prev[0]-curr[0], prev[1]-curr[1])
	#print(data[curr[1]][curr[0]])
	adj = list(deltas[data[curr[1]][curr[0]]])
	if delta not in adj:
		assert "delta not in adj"
	
	#print(adj, delta)
	adj.remove(delta)
	#print(adj, (curr[0]+adj[0][0], curr[1]+adj[0][1]))
	return (curr[0]+adj[0][0], curr[1]+adj[0][1])

loop = None

def part1():
	global loop
	loop = [start]
	
	pos = start
	
	if data[pos[1]-1][pos[0]+0] in ("|", "7", "F"):
		pos = (start[0], start[1]-1)
	elif data[pos[1]+1][pos[0]+0] in ("|", "L", "J"):
		pos = (start[0], start[1]+1)
	elif data[pos[1]+0][pos[0]-1] in ("-", "L", "F"):
		pos = (start[0]-1, start[1])
	elif data[pos[1]+0][pos[0]+1] in ("-", "J", "7"):
		pos = (start[0]+1, start[1])
	else:
		assert "no start"
	
	loop.append(pos)
	
	while pos != start:
		pos = next_pos(loop[-1], loop[-2])
		loop.append(pos)

	assert len(loop) % 2 == 1
	return len(loop) // 2


def bfs(start):
	to_check = [start]
	seen = set()
	
	while len(to_check) > 0:
		#print(len(to_check), len(seen))
		#print(to_check)
		curr = to_check[0]
		to_check.pop(0)
		
		if curr in seen:
			continue
		seen.add(curr)
		
		for adj in ((0, -1), (0, 1), (-1, 0), (1, 0)):
			np = (curr[0]+adj[0], curr[1]+adj[1])
			if 0 <= np[0] < width and 0 <= np[1] < height:
				if np not in loop:
					if np not in seen and np not in to_check:
						to_check.append(np)

	return seen


# https://en.m.wikipedia.org/wiki/Even%E2%80%93odd_rule
def is_point_in_path(x: int, y: int, poly) -> bool:
    """Determine if the point is on the path, corner, or boundary of the polygon

    Args:
      x -- The x coordinates of point.
      y -- The y coordinates of point.
      poly -- a list of tuples [(x, y), (x, y), ...]

    Returns:
      True if the point is in the path or is a corner or on the boundary"""
    num = len(poly)
    j = num - 1
    c = False
    for i in range(num):
        if (x == poly[i][0]) and (y == poly[i][1]):
            # point is a corner
            return True
        if (poly[i][1] > y) != (poly[j][1] > y):
            slope = (x - poly[i][0]) * (poly[j][1] - poly[i][1]) - (
                poly[j][0] - poly[i][0]
            ) * (y - poly[i][1])
            if slope == 0:
                # point is on boundary
                return True
            if (slope < 0) != (poly[j][1] < poly[i][1]):
                c = not c
        j = i
    return c



def check_group(group):
	# check touching edges
	for pos in group:
		if pos[0] == 0 or pos[0] == width-1:
			return False
		if pos[1] == 0 or pos[1] == height-1:
			return False
	
	# group is now surrounded by loop
	
	test_p = list(group)[0]
	
	return is_point_in_path(*test_p, loop)


def part2():
	points = set()
	
	groups = []
	seen = set()
	for x in range(width):
		for y in range(height):
			#print(x, y)
			if (x, y) not in seen and (x, y) not in loop:
				group = bfs((x, y))
				groups.append(group)
				seen |= set(group)
				#print(x, y, len(group))
				
	for g in groups:
		if check_group(g):
			points |= g
	
	return len(points)

p1()
p2()
